EasyReflectometry/sample/items/gradient_layer.py

```diff
@@ -0,0 +1,239 @@
+from __future__ import annotations
+
+from easyCore.Fitting.Constraints import ObjConstraint
+from numpy import arange
+
+from EasyReflectometry.sample.layer import Layer
+from EasyReflectometry.sample.material import Material
+
+from .multilayer import MultiLayer
+
+
+class GradientLayer(MultiLayer):
+    """
+    A :py:class:`GradientLayer` constructs a gradient multilayer for the
+    provided initial and final material.
+    """
+    def __init__(
+        self,
+        initial_material: Material,
+        final_material: Material,
+        thickness: float,
+        roughness: float,
+        discretisation_elements: int = 10,
+        name: str = 'EasyGradienLayer',
+        interface=None
+    ) -> GradientLayer:
+        """
+        :param initial_material: Material of initial "part" of the layer
+        :param final_material: Material of final "part" of the layer
+        :param thickness: Thicknkess of the layer
+        :param roughness: Roughness of the layer
+        :param discretisation_elements: Number of dicrete layers
+        :param name: Name for gradient layer
+        :param interface: Interface to use for the layer
+        """
+        self.initial_material = initial_material
+        self.final_material = final_material
+        self.discretisation_elements = discretisation_elements
+
+        gradient_layers = _prepare_gradient_layers(
+            initial_material=initial_material,
+            final_material=final_material,
+            discretisation_elements=discretisation_elements,
+            interface=interface)
+
+        super().__init__(
+            layers=gradient_layers,
+            name=name,
+            interface=interface,
+            type='Gradient-layer'
+        )
+
+        _apply_thickness_constraints(self.layers)
+        self.thickness = thickness
+        self.roughness = roughness
+
+    @property
+    def thickness(self) -> float:
+        """
+        :return: Thickness of the gradient layer
+        """
+        # Layer 0 is the deciding layer as set in _apply_thickness_constraints
+        return self.layers[0].thickness.raw_value * self.discretisation_elements
+
+    @thickness.setter
+    def thickness(self, thickness: float) -> None:
+        """
+        :param thickness: Thickness of the gradient layer
+        """
+        # Layer 0 is the deciding layer as set in _apply_thickness_constraints
+        self.layers[0].thickness.value = thickness / self.discretisation_elements
```

Comment:
`self.discretisation_elements` can be 0, since it can be set externally (constructor). It would probably be prudent to catch it (with a conditional or an exception)